{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "aware-conflict",
   "metadata": {},
   "source": [
    "# Analysis of differential expression across conditions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "extraordinary-individual",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(\"DESeq2\")\n",
    "setwd(\"~/Documents/Proyecto_rodopsina/\")\n",
    "load(\"Data/DokdoniaTPMCounts.RData\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "loaded-rider",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 5 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Temperature</th><th scope=col>Light</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>D_10_R1.sam</th><td>10</td><td>D</td></tr>\n",
       "\t<tr><th scope=row>D_10_R2.sam</th><td>10</td><td>D</td></tr>\n",
       "\t<tr><th scope=row>D_10_R3.sam</th><td>10</td><td>D</td></tr>\n",
       "\t<tr><th scope=row>D_18_R1.sam</th><td>18</td><td>D</td></tr>\n",
       "\t<tr><th scope=row>D_18_R2.sam</th><td>18</td><td>D</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 5 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & Temperature & Light\\\\\n",
       "  & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\tD\\_10\\_R1.sam & 10 & D\\\\\n",
       "\tD\\_10\\_R2.sam & 10 & D\\\\\n",
       "\tD\\_10\\_R3.sam & 10 & D\\\\\n",
       "\tD\\_18\\_R1.sam & 18 & D\\\\\n",
       "\tD\\_18\\_R2.sam & 18 & D\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 5 × 2\n",
       "\n",
       "| <!--/--> | Temperature &lt;chr&gt; | Light &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| D_10_R1.sam | 10 | D |\n",
       "| D_10_R2.sam | 10 | D |\n",
       "| D_10_R3.sam | 10 | D |\n",
       "| D_18_R1.sam | 18 | D |\n",
       "| D_18_R2.sam | 18 | D |\n",
       "\n"
      ],
      "text/plain": [
       "            Temperature Light\n",
       "D_10_R1.sam 10          D    \n",
       "D_10_R2.sam 10          D    \n",
       "D_10_R3.sam 10          D    \n",
       "D_18_R1.sam 18          D    \n",
       "D_18_R2.sam 18          D    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cnames <- colnames(data$counts)\n",
    "conditions = unlist(\n",
    "    lapply(cnames, function(s){return(strsplit(s, \".sam\")[[1]])})\n",
    ")\n",
    "\n",
    "light_conditions <- conditions[1:16]\n",
    "dark_conditions <- conditions[17:length(conditions)]\n",
    "\n",
    "light <- vector()\n",
    "temp <- vector()\n",
    "files <- colnames(data$counts)\n",
    "for (i in 1:length(files)) {\n",
    "    light <- c(light, substring(files[i], 1, 1))\n",
    "    temp <- c(temp, substring(files[i], 3, 4))\n",
    "}\n",
    "\n",
    "meta <- data.frame(row.names = cnames,\n",
    "                  \"Temperature\" = temp,\n",
    "                  \"Light\" = light)\n",
    "head(meta, 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "improved-carrier",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in DESeqDataSet(se, design = design, ignoreRank):\n",
      "“some variables in design formula are characters, converting to factors”\n"
     ]
    }
   ],
   "source": [
    "dds <- DESeqDataSetFromMatrix(countData = data$counts, \n",
    "                              colData = meta, design = ~ Temperature)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "quantitative-location",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n"
     ]
    }
   ],
   "source": [
    "dds_lrt <- DESeq(dds, test=\"LRT\", reduced = ~ 1) # Likelyhood Ratio Test (multiple conditions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "compressed-chicken",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "log2 fold change (MLE): Temperature 34 vs 10 \n",
       "LRT p-value: '~ Temperature' vs '~ 1' \n",
       "DataFrame with 2886 rows and 6 columns\n",
       "              baseMean log2FoldChange     lfcSE      stat       pvalue\n",
       "             <numeric>      <numeric> <numeric> <numeric>    <numeric>\n",
       "MED134_07389  28698.32       0.941887 0.0757369  179.5902  1.08133e-38\n",
       "MED134_07384   3909.36       1.130010 0.0596776  399.6192  2.67690e-86\n",
       "MED134_07379   6942.78       0.903971 0.0447610  469.1036 2.36562e-101\n",
       "MED134_07374   2984.57       0.364880 0.0643950   51.0036  4.88341e-11\n",
       "MED134_07369   3262.90       0.204430 0.0856888   17.1243  6.66336e-04\n",
       "...                ...            ...       ...       ...          ...\n",
       "MED134_14782   1833.07      -0.405973 0.0907462   31.9619  5.33119e-07\n",
       "MED134_14787   4857.90      -0.332865 0.0398016   99.4119  2.07938e-21\n",
       "MED134_14792   4789.24      -0.937420 0.0681734  202.0328  1.53433e-43\n",
       "MED134_14797   1070.22       0.234741 0.0494182  154.9895  2.20969e-33\n",
       "MED134_14802   6407.24      -1.034763 0.0613018  318.9509  7.86759e-69\n",
       "                     padj\n",
       "                <numeric>\n",
       "MED134_07389  4.92227e-38\n",
       "MED134_07384  3.75026e-85\n",
       "MED134_07379 4.18845e-100\n",
       "MED134_07374  8.58315e-11\n",
       "MED134_07369  8.03277e-04\n",
       "...                   ...\n",
       "MED134_14782  7.59794e-07\n",
       "MED134_14787  5.48044e-21\n",
       "MED134_14792  7.93564e-43\n",
       "MED134_14797  8.64116e-33\n",
       "MED134_14802  7.67090e-68"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "res <- results(dds_lrt, alpha = 1e-5)\n",
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "sized-pocket",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "out of 2886 with nonzero total read count\n",
      "adjusted p-value < 1e-05\n",
      "LFC > 0 (up)       : 1053, 36%\n",
      "LFC < 0 (down)     : 1101, 38%\n",
      "outliers [1]       : 0, 0%\n",
      "low counts [2]     : 0, 0%\n",
      "(mean count < 15)\n",
      "[1] see 'cooksCutoff' argument of ?results\n",
      "[2] see 'independentFiltering' argument of ?results\n",
      "\n"
     ]
    }
   ],
   "source": [
    "summary(res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "surface-volleyball",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table width=\"100%\" summary=\"page for results {DESeq2}\"><tr><td>results {DESeq2}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2>Extract results from a DESeq analysis</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p><code>results</code> extracts a result table from a DESeq analysis giving base means across samples,\n",
       "log2 fold changes, standard errors, test statistics, p-values and adjusted p-values;\n",
       "<code>resultsNames</code> returns the names of the estimated effects (coefficents) of the model;\n",
       "<code>removeResults</code> returns a <code>DESeqDataSet</code> object with results columns removed.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre>\n",
       "results(\n",
       "  object,\n",
       "  contrast,\n",
       "  name,\n",
       "  lfcThreshold = 0,\n",
       "  altHypothesis = c(\"greaterAbs\", \"lessAbs\", \"greater\", \"less\"),\n",
       "  listValues = c(1, -1),\n",
       "  cooksCutoff,\n",
       "  independentFiltering = TRUE,\n",
       "  alpha = 0.1,\n",
       "  filter,\n",
       "  theta,\n",
       "  pAdjustMethod = \"BH\",\n",
       "  filterFun,\n",
       "  format = c(\"DataFrame\", \"GRanges\", \"GRangesList\"),\n",
       "  test,\n",
       "  addMLE = FALSE,\n",
       "  tidy = FALSE,\n",
       "  parallel = FALSE,\n",
       "  BPPARAM = bpparam(),\n",
       "  minmu = 0.5\n",
       ")\n",
       "\n",
       "resultsNames(object)\n",
       "\n",
       "removeResults(object)\n",
       "</pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table summary=\"R argblock\">\n",
       "<tr valign=\"top\"><td><code>object</code></td>\n",
       "<td>\n",
       "<p>a DESeqDataSet, on which one\n",
       "of the following functions has already been called:\n",
       "<code>DESeq</code>, <code>nbinomWaldTest</code>, or <code>nbinomLRT</code></p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>contrast</code></td>\n",
       "<td>\n",
       "<p>this argument specifies what comparison to extract from\n",
       "the <code>object</code> to build a results table. one of either:\n",
       "</p>\n",
       "\n",
       "<ul>\n",
       "<li><p> a character vector with exactly three elements:\n",
       "the name of a factor in the design formula,\n",
       "the name of the numerator level for the fold change,\n",
       "and the name of the denominator level for the fold change\n",
       "(simplest case)\n",
       "</p>\n",
       "</li>\n",
       "<li><p> a list of 2 character vectors: the names of the fold changes\n",
       "for the numerator, and the names of the fold changes\n",
       "for the denominator.\n",
       "these names should be elements of <code>resultsNames(object)</code>.\n",
       "if the list is length 1, a second element is added which is the\n",
       "empty character vector, <code>character()</code>.\n",
       "(more general case, can be to combine interaction terms and main effects)\n",
       "</p>\n",
       "</li>\n",
       "<li><p> a numeric contrast vector with one element\n",
       "for each element in <code>resultsNames(object)</code> (most general case)\n",
       "</p>\n",
       "</li></ul>\n",
       "\n",
       "<p>If specified, the <code>name</code> argument is ignored.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>name</code></td>\n",
       "<td>\n",
       "<p>the name of the individual effect (coefficient) for\n",
       "building a results table. Use this argument rather than <code>contrast</code>\n",
       "for continuous variables, individual effects or for individual interaction terms.\n",
       "The value provided to <code>name</code> must be an element of <code>resultsNames(object)</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>lfcThreshold</code></td>\n",
       "<td>\n",
       "<p>a non-negative value which specifies a log2 fold change\n",
       "threshold. The default value is 0, corresponding to a test that\n",
       "the log2 fold changes are equal to zero. The user can\n",
       "specify the alternative hypothesis using the <code>altHypothesis</code> argument,\n",
       "which defaults to testing\n",
       "for log2 fold changes greater in absolute value than a given threshold.\n",
       "If <code>lfcThreshold</code> is specified,\n",
       "the results are for Wald tests, and LRT p-values will be overwritten.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>altHypothesis</code></td>\n",
       "<td>\n",
       "<p>character which specifies the alternative hypothesis,\n",
       "i.e. those values of log2 fold change which the user is interested in\n",
       "finding. The complement of this set of values is the null hypothesis which\n",
       "will be tested. If the log2 fold change specified by <code>name</code>\n",
       "or by <code>contrast</code> is written as <i> beta </i>, then the possible values for\n",
       "<code>altHypothesis</code> represent the following alternate hypotheses:\n",
       "</p>\n",
       "\n",
       "<ul>\n",
       "<li><p> greaterAbs: <i> |beta| &gt; lfcThreshold </i>,\n",
       "and p-values are two-tailed\n",
       "</p>\n",
       "</li>\n",
       "<li><p> lessAbs: <i> |beta| &lt; lfcThreshold </i>,\n",
       "p-values are the maximum of the upper and lower tests.\n",
       "The Wald statistic given is positive, an SE-scaled distance from the closest boundary\n",
       "</p>\n",
       "</li>\n",
       "<li><p> greater: <i> beta &gt; lfcThreshold </i>\n",
       "</p>\n",
       "</li>\n",
       "<li><p> less: <i> beta &lt; -lfcThreshold </i>\n",
       "</p>\n",
       "</li></ul>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>listValues</code></td>\n",
       "<td>\n",
       "<p>only used if a list is provided to <code>contrast</code>:\n",
       "a numeric of length two: the log2 fold changes in the list are multiplied by these values.\n",
       "the first number should be positive and the second negative. \n",
       "by default this is <code>c(1,-1)</code></p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>cooksCutoff</code></td>\n",
       "<td>\n",
       "<p>theshold on Cook's distance, such that if one or more\n",
       "samples for a row have a distance higher, the p-value for the row is\n",
       "set to NA. The default cutoff is the .99 quantile of the F(p, m-p) distribution,\n",
       "where p is the number of coefficients being fitted and m is the number of samples.\n",
       "Set to <code>Inf</code> or <code>FALSE</code> to disable the resetting of p-values to NA.\n",
       "Note: this test excludes the Cook's distance of samples belonging to experimental\n",
       "groups with only 2 samples.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>independentFiltering</code></td>\n",
       "<td>\n",
       "<p>logical, whether independent filtering should be\n",
       "applied automatically</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>alpha</code></td>\n",
       "<td>\n",
       "<p>the significance cutoff used for optimizing the independent\n",
       "filtering (by default 0.1). If the adjusted p-value cutoff (FDR) will be a\n",
       "value other than 0.1, <code>alpha</code> should be set to that value.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>filter</code></td>\n",
       "<td>\n",
       "<p>the vector of filter statistics over which the independent\n",
       "filtering will be optimized. By default the mean of normalized counts is used.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>theta</code></td>\n",
       "<td>\n",
       "<p>the quantiles at which to assess the number of rejections\n",
       "from independent filtering</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>pAdjustMethod</code></td>\n",
       "<td>\n",
       "<p>the method to use for adjusting p-values, see <code>?p.adjust</code></p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>filterFun</code></td>\n",
       "<td>\n",
       "<p>an optional custom function for performing independent filtering\n",
       "and p-value adjustment, with arguments <code>res</code> (a DESeqResults object),\n",
       "<code>filter</code> (the quantitity for filtering tests),\n",
       "<code>alpha</code> (the target FDR),\n",
       "<code>pAdjustMethod</code>. This function should return a DESeqResults object\n",
       "with a <code>padj</code> column.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>format</code></td>\n",
       "<td>\n",
       "<p>character, either <code>\"DataFrame\"</code>,\n",
       "<code>\"GRanges\"</code>, or <code>\"GRangesList\"</code>,\n",
       "whether the results should be printed as a <code>DESeqResults</code> DataFrame,\n",
       "or if the results DataFrame should be attached as metadata columns to\n",
       "the <code>GRanges</code> or <code>GRangesList</code> <code>rowRanges</code> of the <code>DESeqDataSet</code>.\n",
       "If the <code>rowRanges</code> is a <code>GRangesList</code>, and <code>GRanges</code> is requested, \n",
       "the range of each gene will be returned</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>test</code></td>\n",
       "<td>\n",
       "<p>this is automatically detected internally if not provided.\n",
       "the one exception is after <code>nbinomLRT</code> has been run, <code>test=\"Wald\"</code>\n",
       "will generate Wald statistics and Wald test p-values.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>addMLE</code></td>\n",
       "<td>\n",
       "<p>if <code>betaPrior=TRUE</code> was used (non-default),\n",
       "this logical argument specifies if the &quot;unshrunken&quot; maximum likelihood estimates (MLE)\n",
       "of log2 fold change should be added as a column to the results table (default is FALSE).\n",
       "This argument is preserved for backward compatability, as now <code>betaPrior=TRUE</code>\n",
       "by default and the recommended pipeline is\n",
       "to generate shrunken MAP estimates using <code>lfcShrink</code>.\n",
       "This argument functionality is only implemented for <code>contrast</code>\n",
       "specified as three element character vectors.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>tidy</code></td>\n",
       "<td>\n",
       "<p>whether to output the results table with rownames as a first column 'row'.\n",
       "the table will also be coerced to <code>data.frame</code></p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>parallel</code></td>\n",
       "<td>\n",
       "<p>if FALSE, no parallelization. if TRUE, parallel\n",
       "execution using <code>BiocParallel</code>, see next argument <code>BPPARAM</code></p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>BPPARAM</code></td>\n",
       "<td>\n",
       "<p>an optional parameter object passed internally\n",
       "to <code>bplapply</code> when <code>parallel=TRUE</code>.\n",
       "If not specified, the parameters last registered with\n",
       "<code>register</code> will be used.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>minmu</code></td>\n",
       "<td>\n",
       "<p>lower bound on the estimated count (used when calculating contrasts)</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Details</h3>\n",
       "\n",
       "<p>The results table when printed will provide the information about\n",
       "the comparison, e.g. &quot;log2 fold change (MAP): condition treated vs untreated&quot;, meaning\n",
       "that the estimates are of log2(treated / untreated), as would be returned by\n",
       "<code>contrast=c(\"condition\",\"treated\",\"untreated\")</code>.\n",
       "Multiple results can be returned for analyses beyond a simple two group comparison,\n",
       "so <code>results</code> takes arguments <code>contrast</code> and <code>name</code> to help\n",
       "the user pick out the comparisons of interest for printing a results table.\n",
       "The use of the <code>contrast</code> argument is recommended for exact specification\n",
       "of the levels which should be compared and their order.\n",
       "</p>\n",
       "<p>If <code>results</code> is run without specifying <code>contrast</code> or <code>name</code>,\n",
       "it will return the comparison of the last level of the last variable in the\n",
       "design formula over the first level of this variable. For example, for a simple two-group\n",
       "comparison, this would return the log2 fold changes of the second group over the\n",
       "first group (the reference level). Please see examples below and in the vignette. \n",
       "</p>\n",
       "<p>The argument <code>contrast</code> can be used to generate results tables for\n",
       "any comparison of interest, for example, the log2 fold change between\n",
       "two levels of a factor, and its usage is described below. It can also\n",
       "accomodate more complicated numeric comparisons.\n",
       "Note that <code>contrast</code> will set to 0 the estimated LFC in a\n",
       "comparison of two groups, where all of the counts in the two groups\n",
       "are equal to 0 (while other groups have positive counts), while\n",
       "<code>name</code> will not automatically set these LFC to 0.\n",
       "The test statistic used for a contrast is:\n",
       "</p>\n",
       "<p style=\"text-align: center;\"><i> c' beta / sqrt( c' Sigma c ) </i></p>\n",
       "\n",
       "<p>The argument <code>name</code> can be used to generate results tables for\n",
       "individual effects, which must be individual elements of <code>resultsNames(object)</code>.\n",
       "These individual effects could represent continuous covariates, effects\n",
       "for individual levels, or individual interaction effects.\n",
       "</p>\n",
       "<p>Information on the comparison which was used to build the results table,\n",
       "and the statistical test which was used for p-values (Wald test or likelihood ratio test)\n",
       "is stored within the object returned by <code>results</code>. This information is in\n",
       "the metadata columns of the results table, which is accessible by calling <code>mcols</code>\n",
       "on the <code>DESeqResults</code> object returned by <code>results</code>.\n",
       "</p>\n",
       "<p>On p-values:\n",
       "</p>\n",
       "<p>By default, independent filtering is performed to select a set of genes\n",
       "for multiple test correction which maximizes the number of adjusted\n",
       "p-values less than a given critical value <code>alpha</code> (by default 0.1).\n",
       "See the reference in this man page for details on independent filtering.\n",
       "The filter used for maximizing the number of rejections is the mean\n",
       "of normalized counts for all samples in the dataset.\n",
       "Several arguments from the <code>filtered_p</code> function of\n",
       "the genefilter package (used within the <code>results</code> function)\n",
       "are provided here to control the independent filtering behavior.\n",
       "In DESeq2 version &gt;= 1.10, the threshold that is chosen is\n",
       "the lowest quantile of the filter for which the\n",
       "number of rejections is close to the peak of a curve fit\n",
       "to the number of rejections over the filter quantiles.\n",
       "'Close to' is defined as within 1 residual standard deviation.\n",
       "The adjusted p-values for the genes which do not pass the filter threshold\n",
       "are set to <code>NA</code>. \n",
       "</p>\n",
       "<p>By default, <code>results</code> assigns a p-value of <code>NA</code>\n",
       "to genes containing count outliers, as identified using Cook's distance.\n",
       "See the <code>cooksCutoff</code> argument for control of this behavior.\n",
       "Cook's distances for each sample are accessible as a matrix &quot;cooks&quot;\n",
       "stored in the <code>assays()</code> list. This measure is useful for identifying rows where the\n",
       "observed counts might not fit to a Negative Binomial distribution.\n",
       "</p>\n",
       "<p>For analyses using the likelihood ratio test (using <code>nbinomLRT</code>),\n",
       "the p-values are determined solely by the difference in deviance between\n",
       "the full and reduced model formula. A single log2 fold change is printed\n",
       "in the results table for consistency with other results table outputs,\n",
       "however the test statistic and p-values may nevertheless involve\n",
       "the testing of one or more log2 fold changes.\n",
       "Which log2 fold change is printed in the results table can be controlled\n",
       "using the <code>name</code> argument, or by default this will be the estimated\n",
       "coefficient for the last element of <code>resultsNames(object)</code>.\n",
       "</p>\n",
       "<p>If <code>useT=TRUE</code> was specified when running <code>DESeq</code> or <code>nbinomWaldTest</code>,\n",
       "then the p-value generated by <code>results</code> will also make use of the\n",
       "t distribution for the Wald statistic, using the degrees of freedom\n",
       "in <code>mcols(object)$tDegreesFreedom</code>.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Value</h3>\n",
       "\n",
       "<p>For <code>results</code>: a <code>DESeqResults</code> object, which is\n",
       "a simple subclass of DataFrame. This object contains the results columns:\n",
       "<code>baseMean</code>, <code>log2FoldChange</code>, <code>lfcSE</code>, <code>stat</code>,\n",
       "<code>pvalue</code> and <code>padj</code>,\n",
       "and also includes metadata columns of variable information.\n",
       "The <code>lfcSE</code> gives the standard error of the <code>log2FoldChange</code>.\n",
       "For the Wald test, <code>stat</code> is the Wald statistic: the <code>log2FoldChange</code>\n",
       "divided by <code>lfcSE</code>, which is compared to a standard Normal distribution\n",
       "to generate a two-tailed <code>pvalue</code>. For the likelihood ratio test (LRT),\n",
       "<code>stat</code> is the difference in deviance between the reduced model and the full model,\n",
       "which is compared to a chi-squared distribution to generate a <code>pvalue</code>.\n",
       "</p>\n",
       "<p>For <code>resultsNames</code>: the names of the columns available as results,\n",
       "usually a combination of the variable name and a level\n",
       "</p>\n",
       "<p>For <code>removeResults</code>: the original <code>DESeqDataSet</code> with results metadata columns removed\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>References</h3>\n",
       "\n",
       "<p>Richard Bourgon, Robert Gentleman, Wolfgang Huber: Independent\n",
       "filtering increases detection power for high-throughput experiments.\n",
       "PNAS (2010), <a href=\"http://dx.doi.org/10.1073/pnas.0914005107\">http://dx.doi.org/10.1073/pnas.0914005107</a>\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>See Also</h3>\n",
       "\n",
       "<p><code>DESeq</code>, <code>lfcShrink</code>\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre>\n",
       "\n",
       "## Example 1: two-group comparison\n",
       "\n",
       "dds &lt;- makeExampleDESeqDataSet(m=4)\n",
       "\n",
       "dds &lt;- DESeq(dds)\n",
       "res &lt;- results(dds, contrast=c(\"condition\",\"B\",\"A\"))\n",
       "\n",
       "# with more than two groups, the call would look similar, e.g.:\n",
       "# results(dds, contrast=c(\"condition\",\"C\",\"A\"))\n",
       "# etc.\n",
       "\n",
       "## Example 2: two conditions, two genotypes, with an interaction term\n",
       "\n",
       "dds &lt;- makeExampleDESeqDataSet(n=100,m=12)\n",
       "dds$genotype &lt;- factor(rep(rep(c(\"I\",\"II\"),each=3),2))\n",
       "\n",
       "design(dds) &lt;- ~ genotype + condition + genotype:condition\n",
       "dds &lt;- DESeq(dds) \n",
       "resultsNames(dds)\n",
       "\n",
       "# the condition effect for genotype I (the main effect)\n",
       "results(dds, contrast=c(\"condition\",\"B\",\"A\"))\n",
       "\n",
       "# the condition effect for genotype II\n",
       "# this is, by definition, the main effect *plus* the interaction term\n",
       "# (the extra condition effect in genotype II compared to genotype I).\n",
       "results(dds, list( c(\"condition_B_vs_A\",\"genotypeII.conditionB\") ))\n",
       "\n",
       "# the interaction term, answering: is the condition effect *different* across genotypes?\n",
       "results(dds, name=\"genotypeII.conditionB\")\n",
       " \n",
       "## Example 3: two conditions, three genotypes\n",
       "\n",
       "# ~~~ Using interaction terms ~~~\n",
       "\n",
       "dds &lt;- makeExampleDESeqDataSet(n=100,m=18)\n",
       "dds$genotype &lt;- factor(rep(rep(c(\"I\",\"II\",\"III\"),each=3),2))\n",
       "design(dds) &lt;- ~ genotype + condition + genotype:condition\n",
       "dds &lt;- DESeq(dds)\n",
       "resultsNames(dds)\n",
       "\n",
       "# the condition effect for genotype I (the main effect)\n",
       "results(dds, contrast=c(\"condition\",\"B\",\"A\"))\n",
       "\n",
       "# the condition effect for genotype III.\n",
       "# this is the main effect *plus* the interaction term\n",
       "# (the extra condition effect in genotype III compared to genotype I).\n",
       "results(dds, contrast=list( c(\"condition_B_vs_A\",\"genotypeIII.conditionB\") ))\n",
       " \n",
       "# the interaction term for condition effect in genotype III vs genotype I.\n",
       "# this tests if the condition effect is different in III compared to I\n",
       "results(dds, name=\"genotypeIII.conditionB\")\n",
       "\n",
       "# the interaction term for condition effect in genotype III vs genotype II.\n",
       "# this tests if the condition effect is different in III compared to II\n",
       "results(dds, contrast=list(\"genotypeIII.conditionB\", \"genotypeII.conditionB\"))\n",
       "\n",
       "# Note that a likelihood ratio could be used to test if there are any\n",
       "# differences in the condition effect between the three genotypes.\n",
       "\n",
       "# ~~~ Using a grouping variable ~~~\n",
       "\n",
       "# This is a useful construction when users just want to compare\n",
       "# specific groups which are combinations of variables.\n",
       "\n",
       "dds$group &lt;- factor(paste0(dds$genotype, dds$condition))\n",
       "design(dds) &lt;- ~ group\n",
       "dds &lt;- DESeq(dds)\n",
       "resultsNames(dds)\n",
       "\n",
       "# the condition effect for genotypeIII\n",
       "results(dds, contrast=c(\"group\", \"IIIB\", \"IIIA\"))\n",
       "\n",
       "</pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>DESeq2</em> version 1.30.1 ]</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{results}{Extract results from a DESeq analysis}{results}\n",
       "\\aliasA{removeResults}{results}{removeResults}\n",
       "\\aliasA{resultsNames}{results}{resultsNames}\n",
       "%\n",
       "\\begin{Description}\\relax\n",
       "\\code{results} extracts a result table from a DESeq analysis giving base means across samples,\n",
       "log2 fold changes, standard errors, test statistics, p-values and adjusted p-values;\n",
       "\\code{resultsNames} returns the names of the estimated effects (coefficents) of the model;\n",
       "\\code{removeResults} returns a \\code{DESeqDataSet} object with results columns removed.\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "results(\n",
       "  object,\n",
       "  contrast,\n",
       "  name,\n",
       "  lfcThreshold = 0,\n",
       "  altHypothesis = c(\"greaterAbs\", \"lessAbs\", \"greater\", \"less\"),\n",
       "  listValues = c(1, -1),\n",
       "  cooksCutoff,\n",
       "  independentFiltering = TRUE,\n",
       "  alpha = 0.1,\n",
       "  filter,\n",
       "  theta,\n",
       "  pAdjustMethod = \"BH\",\n",
       "  filterFun,\n",
       "  format = c(\"DataFrame\", \"GRanges\", \"GRangesList\"),\n",
       "  test,\n",
       "  addMLE = FALSE,\n",
       "  tidy = FALSE,\n",
       "  parallel = FALSE,\n",
       "  BPPARAM = bpparam(),\n",
       "  minmu = 0.5\n",
       ")\n",
       "\n",
       "resultsNames(object)\n",
       "\n",
       "removeResults(object)\n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{object}] a DESeqDataSet, on which one\n",
       "of the following functions has already been called:\n",
       "\\code{\\LinkA{DESeq}{DESeq}}, \\code{\\LinkA{nbinomWaldTest}{nbinomWaldTest}}, or \\code{\\LinkA{nbinomLRT}{nbinomLRT}}\n",
       "\n",
       "\\item[\\code{contrast}] this argument specifies what comparison to extract from\n",
       "the \\code{object} to build a results table. one of either:\n",
       "\\begin{itemize}\n",
       "\n",
       "\\item{} a character vector with exactly three elements:\n",
       "the name of a factor in the design formula,\n",
       "the name of the numerator level for the fold change,\n",
       "and the name of the denominator level for the fold change\n",
       "(simplest case)\n",
       "\\item{} a list of 2 character vectors: the names of the fold changes\n",
       "for the numerator, and the names of the fold changes\n",
       "for the denominator.\n",
       "these names should be elements of \\code{resultsNames(object)}.\n",
       "if the list is length 1, a second element is added which is the\n",
       "empty character vector, \\code{character()}.\n",
       "(more general case, can be to combine interaction terms and main effects)\n",
       "\\item{} a numeric contrast vector with one element\n",
       "for each element in \\code{resultsNames(object)} (most general case)\n",
       "\n",
       "\\end{itemize}\n",
       "\n",
       "If specified, the \\code{name} argument is ignored.\n",
       "\n",
       "\\item[\\code{name}] the name of the individual effect (coefficient) for\n",
       "building a results table. Use this argument rather than \\code{contrast}\n",
       "for continuous variables, individual effects or for individual interaction terms.\n",
       "The value provided to \\code{name} must be an element of \\code{resultsNames(object)}.\n",
       "\n",
       "\\item[\\code{lfcThreshold}] a non-negative value which specifies a log2 fold change\n",
       "threshold. The default value is 0, corresponding to a test that\n",
       "the log2 fold changes are equal to zero. The user can\n",
       "specify the alternative hypothesis using the \\code{altHypothesis} argument,\n",
       "which defaults to testing\n",
       "for log2 fold changes greater in absolute value than a given threshold.\n",
       "If \\code{lfcThreshold} is specified,\n",
       "the results are for Wald tests, and LRT p-values will be overwritten.\n",
       "\n",
       "\\item[\\code{altHypothesis}] character which specifies the alternative hypothesis,\n",
       "i.e. those values of log2 fold change which the user is interested in\n",
       "finding. The complement of this set of values is the null hypothesis which\n",
       "will be tested. If the log2 fold change specified by \\code{name}\n",
       "or by \\code{contrast} is written as \\eqn{ \\beta }{}, then the possible values for\n",
       "\\code{altHypothesis} represent the following alternate hypotheses:\n",
       "\\begin{itemize}\n",
       "\n",
       "\\item{} greaterAbs: \\eqn{|\\beta| > \\textrm{lfcThreshold} }{},\n",
       "and p-values are two-tailed\n",
       "\\item{} lessAbs: \\eqn{ |\\beta| < \\textrm{lfcThreshold} }{},\n",
       "p-values are the maximum of the upper and lower tests.\n",
       "The Wald statistic given is positive, an SE-scaled distance from the closest boundary\n",
       "\\item{} greater: \\eqn{ \\beta > \\textrm{lfcThreshold} }{}\n",
       "\\item{} less: \\eqn{ \\beta < -\\textrm{lfcThreshold} }{}\n",
       "\n",
       "\\end{itemize}\n",
       "\n",
       "\n",
       "\\item[\\code{listValues}] only used if a list is provided to \\code{contrast}:\n",
       "a numeric of length two: the log2 fold changes in the list are multiplied by these values.\n",
       "the first number should be positive and the second negative. \n",
       "by default this is \\code{c(1,-1)}\n",
       "\n",
       "\\item[\\code{cooksCutoff}] theshold on Cook's distance, such that if one or more\n",
       "samples for a row have a distance higher, the p-value for the row is\n",
       "set to NA. The default cutoff is the .99 quantile of the F(p, m-p) distribution,\n",
       "where p is the number of coefficients being fitted and m is the number of samples.\n",
       "Set to \\code{Inf} or \\code{FALSE} to disable the resetting of p-values to NA.\n",
       "Note: this test excludes the Cook's distance of samples belonging to experimental\n",
       "groups with only 2 samples.\n",
       "\n",
       "\\item[\\code{independentFiltering}] logical, whether independent filtering should be\n",
       "applied automatically\n",
       "\n",
       "\\item[\\code{alpha}] the significance cutoff used for optimizing the independent\n",
       "filtering (by default 0.1). If the adjusted p-value cutoff (FDR) will be a\n",
       "value other than 0.1, \\code{alpha} should be set to that value.\n",
       "\n",
       "\\item[\\code{filter}] the vector of filter statistics over which the independent\n",
       "filtering will be optimized. By default the mean of normalized counts is used.\n",
       "\n",
       "\\item[\\code{theta}] the quantiles at which to assess the number of rejections\n",
       "from independent filtering\n",
       "\n",
       "\\item[\\code{pAdjustMethod}] the method to use for adjusting p-values, see \\code{?p.adjust}\n",
       "\n",
       "\\item[\\code{filterFun}] an optional custom function for performing independent filtering\n",
       "and p-value adjustment, with arguments \\code{res} (a DESeqResults object),\n",
       "\\code{filter} (the quantitity for filtering tests),\n",
       "\\code{alpha} (the target FDR),\n",
       "\\code{pAdjustMethod}. This function should return a DESeqResults object\n",
       "with a \\code{padj} column.\n",
       "\n",
       "\\item[\\code{format}] character, either \\code{\"DataFrame\"},\n",
       "\\code{\"GRanges\"}, or \\code{\"GRangesList\"},\n",
       "whether the results should be printed as a \\code{\\LinkA{DESeqResults}{DESeqResults}} DataFrame,\n",
       "or if the results DataFrame should be attached as metadata columns to\n",
       "the \\code{GRanges} or \\code{GRangesList} \\code{rowRanges} of the \\code{DESeqDataSet}.\n",
       "If the \\code{rowRanges} is a \\code{GRangesList}, and \\code{GRanges} is requested, \n",
       "the range of each gene will be returned\n",
       "\n",
       "\\item[\\code{test}] this is automatically detected internally if not provided.\n",
       "the one exception is after \\code{nbinomLRT} has been run, \\code{test=\"Wald\"}\n",
       "will generate Wald statistics and Wald test p-values.\n",
       "\n",
       "\\item[\\code{addMLE}] if \\code{betaPrior=TRUE} was used (non-default),\n",
       "this logical argument specifies if the \"unshrunken\" maximum likelihood estimates (MLE)\n",
       "of log2 fold change should be added as a column to the results table (default is FALSE).\n",
       "This argument is preserved for backward compatability, as now \\code{betaPrior=TRUE}\n",
       "by default and the recommended pipeline is\n",
       "to generate shrunken MAP estimates using \\code{\\LinkA{lfcShrink}{lfcShrink}}.\n",
       "This argument functionality is only implemented for \\code{contrast}\n",
       "specified as three element character vectors.\n",
       "\n",
       "\\item[\\code{tidy}] whether to output the results table with rownames as a first column 'row'.\n",
       "the table will also be coerced to \\code{data.frame}\n",
       "\n",
       "\\item[\\code{parallel}] if FALSE, no parallelization. if TRUE, parallel\n",
       "execution using \\code{BiocParallel}, see next argument \\code{BPPARAM}\n",
       "\n",
       "\\item[\\code{BPPARAM}] an optional parameter object passed internally\n",
       "to \\code{\\LinkA{bplapply}{bplapply}} when \\code{parallel=TRUE}.\n",
       "If not specified, the parameters last registered with\n",
       "\\code{\\LinkA{register}{register}} will be used.\n",
       "\n",
       "\\item[\\code{minmu}] lower bound on the estimated count (used when calculating contrasts)\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Details}\\relax\n",
       "The results table when printed will provide the information about\n",
       "the comparison, e.g. \"log2 fold change (MAP): condition treated vs untreated\", meaning\n",
       "that the estimates are of log2(treated / untreated), as would be returned by\n",
       "\\code{contrast=c(\"condition\",\"treated\",\"untreated\")}.\n",
       "Multiple results can be returned for analyses beyond a simple two group comparison,\n",
       "so \\code{results} takes arguments \\code{contrast} and \\code{name} to help\n",
       "the user pick out the comparisons of interest for printing a results table.\n",
       "The use of the \\code{contrast} argument is recommended for exact specification\n",
       "of the levels which should be compared and their order.\n",
       "\n",
       "If \\code{results} is run without specifying \\code{contrast} or \\code{name},\n",
       "it will return the comparison of the last level of the last variable in the\n",
       "design formula over the first level of this variable. For example, for a simple two-group\n",
       "comparison, this would return the log2 fold changes of the second group over the\n",
       "first group (the reference level). Please see examples below and in the vignette. \n",
       "\n",
       "The argument \\code{contrast} can be used to generate results tables for\n",
       "any comparison of interest, for example, the log2 fold change between\n",
       "two levels of a factor, and its usage is described below. It can also\n",
       "accomodate more complicated numeric comparisons.\n",
       "Note that \\code{contrast} will set to 0 the estimated LFC in a\n",
       "comparison of two groups, where all of the counts in the two groups\n",
       "are equal to 0 (while other groups have positive counts), while\n",
       "\\code{name} will not automatically set these LFC to 0.\n",
       "The test statistic used for a contrast is:\n",
       "\n",
       "\\deqn{ c^t \\beta / \\sqrt{c^t \\Sigma c } }{}\n",
       "\n",
       "The argument \\code{name} can be used to generate results tables for\n",
       "individual effects, which must be individual elements of \\code{resultsNames(object)}.\n",
       "These individual effects could represent continuous covariates, effects\n",
       "for individual levels, or individual interaction effects.\n",
       "\n",
       "Information on the comparison which was used to build the results table,\n",
       "and the statistical test which was used for p-values (Wald test or likelihood ratio test)\n",
       "is stored within the object returned by \\code{results}. This information is in\n",
       "the metadata columns of the results table, which is accessible by calling \\code{mcols}\n",
       "on the \\code{\\LinkA{DESeqResults}{DESeqResults}} object returned by \\code{results}.\n",
       "\n",
       "On p-values:\n",
       "\n",
       "By default, independent filtering is performed to select a set of genes\n",
       "for multiple test correction which maximizes the number of adjusted\n",
       "p-values less than a given critical value \\code{alpha} (by default 0.1).\n",
       "See the reference in this man page for details on independent filtering.\n",
       "The filter used for maximizing the number of rejections is the mean\n",
       "of normalized counts for all samples in the dataset.\n",
       "Several arguments from the \\code{filtered\\_p} function of\n",
       "the genefilter package (used within the \\code{results} function)\n",
       "are provided here to control the independent filtering behavior.\n",
       "In DESeq2 version >= 1.10, the threshold that is chosen is\n",
       "the lowest quantile of the filter for which the\n",
       "number of rejections is close to the peak of a curve fit\n",
       "to the number of rejections over the filter quantiles.\n",
       "'Close to' is defined as within 1 residual standard deviation.\n",
       "The adjusted p-values for the genes which do not pass the filter threshold\n",
       "are set to \\code{NA}. \n",
       "\n",
       "By default, \\code{results} assigns a p-value of \\code{NA}\n",
       "to genes containing count outliers, as identified using Cook's distance.\n",
       "See the \\code{cooksCutoff} argument for control of this behavior.\n",
       "Cook's distances for each sample are accessible as a matrix \"cooks\"\n",
       "stored in the \\code{assays()} list. This measure is useful for identifying rows where the\n",
       "observed counts might not fit to a Negative Binomial distribution.\n",
       "\n",
       "For analyses using the likelihood ratio test (using \\code{\\LinkA{nbinomLRT}{nbinomLRT}}),\n",
       "the p-values are determined solely by the difference in deviance between\n",
       "the full and reduced model formula. A single log2 fold change is printed\n",
       "in the results table for consistency with other results table outputs,\n",
       "however the test statistic and p-values may nevertheless involve\n",
       "the testing of one or more log2 fold changes.\n",
       "Which log2 fold change is printed in the results table can be controlled\n",
       "using the \\code{name} argument, or by default this will be the estimated\n",
       "coefficient for the last element of \\code{resultsNames(object)}.\n",
       "\n",
       "If \\code{useT=TRUE} was specified when running \\code{DESeq} or \\code{nbinomWaldTest},\n",
       "then the p-value generated by \\code{results} will also make use of the\n",
       "t distribution for the Wald statistic, using the degrees of freedom\n",
       "in \\code{mcols(object)\\$tDegreesFreedom}.\n",
       "\\end{Details}\n",
       "%\n",
       "\\begin{Value}\n",
       "For \\code{results}: a \\code{\\LinkA{DESeqResults}{DESeqResults}} object, which is\n",
       "a simple subclass of DataFrame. This object contains the results columns:\n",
       "\\code{baseMean}, \\code{log2FoldChange}, \\code{lfcSE}, \\code{stat},\n",
       "\\code{pvalue} and \\code{padj},\n",
       "and also includes metadata columns of variable information.\n",
       "The \\code{lfcSE} gives the standard error of the \\code{log2FoldChange}.\n",
       "For the Wald test, \\code{stat} is the Wald statistic: the \\code{log2FoldChange}\n",
       "divided by \\code{lfcSE}, which is compared to a standard Normal distribution\n",
       "to generate a two-tailed \\code{pvalue}. For the likelihood ratio test (LRT),\n",
       "\\code{stat} is the difference in deviance between the reduced model and the full model,\n",
       "which is compared to a chi-squared distribution to generate a \\code{pvalue}.\n",
       "\n",
       "For \\code{resultsNames}: the names of the columns available as results,\n",
       "usually a combination of the variable name and a level\n",
       "\n",
       "For \\code{removeResults}: the original \\code{DESeqDataSet} with results metadata columns removed\n",
       "\\end{Value}\n",
       "%\n",
       "\\begin{References}\\relax\n",
       "Richard Bourgon, Robert Gentleman, Wolfgang Huber: Independent\n",
       "filtering increases detection power for high-throughput experiments.\n",
       "PNAS (2010), \\url{http://dx.doi.org/10.1073/pnas.0914005107}\n",
       "\\end{References}\n",
       "%\n",
       "\\begin{SeeAlso}\\relax\n",
       "\\code{\\LinkA{DESeq}{DESeq}}, \\code{\\LinkA{lfcShrink}{lfcShrink}}\n",
       "\\end{SeeAlso}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "\n",
       "## Example 1: two-group comparison\n",
       "\n",
       "dds <- makeExampleDESeqDataSet(m=4)\n",
       "\n",
       "dds <- DESeq(dds)\n",
       "res <- results(dds, contrast=c(\"condition\",\"B\",\"A\"))\n",
       "\n",
       "# with more than two groups, the call would look similar, e.g.:\n",
       "# results(dds, contrast=c(\"condition\",\"C\",\"A\"))\n",
       "# etc.\n",
       "\n",
       "## Example 2: two conditions, two genotypes, with an interaction term\n",
       "\n",
       "dds <- makeExampleDESeqDataSet(n=100,m=12)\n",
       "dds$genotype <- factor(rep(rep(c(\"I\",\"II\"),each=3),2))\n",
       "\n",
       "design(dds) <- ~ genotype + condition + genotype:condition\n",
       "dds <- DESeq(dds) \n",
       "resultsNames(dds)\n",
       "\n",
       "# the condition effect for genotype I (the main effect)\n",
       "results(dds, contrast=c(\"condition\",\"B\",\"A\"))\n",
       "\n",
       "# the condition effect for genotype II\n",
       "# this is, by definition, the main effect *plus* the interaction term\n",
       "# (the extra condition effect in genotype II compared to genotype I).\n",
       "results(dds, list( c(\"condition_B_vs_A\",\"genotypeII.conditionB\") ))\n",
       "\n",
       "# the interaction term, answering: is the condition effect *different* across genotypes?\n",
       "results(dds, name=\"genotypeII.conditionB\")\n",
       " \n",
       "## Example 3: two conditions, three genotypes\n",
       "\n",
       "# ~~~ Using interaction terms ~~~\n",
       "\n",
       "dds <- makeExampleDESeqDataSet(n=100,m=18)\n",
       "dds$genotype <- factor(rep(rep(c(\"I\",\"II\",\"III\"),each=3),2))\n",
       "design(dds) <- ~ genotype + condition + genotype:condition\n",
       "dds <- DESeq(dds)\n",
       "resultsNames(dds)\n",
       "\n",
       "# the condition effect for genotype I (the main effect)\n",
       "results(dds, contrast=c(\"condition\",\"B\",\"A\"))\n",
       "\n",
       "# the condition effect for genotype III.\n",
       "# this is the main effect *plus* the interaction term\n",
       "# (the extra condition effect in genotype III compared to genotype I).\n",
       "results(dds, contrast=list( c(\"condition_B_vs_A\",\"genotypeIII.conditionB\") ))\n",
       " \n",
       "# the interaction term for condition effect in genotype III vs genotype I.\n",
       "# this tests if the condition effect is different in III compared to I\n",
       "results(dds, name=\"genotypeIII.conditionB\")\n",
       "\n",
       "# the interaction term for condition effect in genotype III vs genotype II.\n",
       "# this tests if the condition effect is different in III compared to II\n",
       "results(dds, contrast=list(\"genotypeIII.conditionB\", \"genotypeII.conditionB\"))\n",
       "\n",
       "# Note that a likelihood ratio could be used to test if there are any\n",
       "# differences in the condition effect between the three genotypes.\n",
       "\n",
       "# ~~~ Using a grouping variable ~~~\n",
       "\n",
       "# This is a useful construction when users just want to compare\n",
       "# specific groups which are combinations of variables.\n",
       "\n",
       "dds$group <- factor(paste0(dds$genotype, dds$condition))\n",
       "design(dds) <- ~ group\n",
       "dds <- DESeq(dds)\n",
       "resultsNames(dds)\n",
       "\n",
       "# the condition effect for genotypeIII\n",
       "results(dds, contrast=c(\"group\", \"IIIB\", \"IIIA\"))\n",
       "\n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "results                 package:DESeq2                 R Documentation\n",
       "\n",
       "_\bE_\bx_\bt_\br_\ba_\bc_\bt _\br_\be_\bs_\bu_\bl_\bt_\bs _\bf_\br_\bo_\bm _\ba _\bD_\bE_\bS_\be_\bq _\ba_\bn_\ba_\bl_\by_\bs_\bi_\bs\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     ‘results’ extracts a result table from a DESeq analysis giving\n",
       "     base means across samples, log2 fold changes, standard errors,\n",
       "     test statistics, p-values and adjusted p-values; ‘resultsNames’\n",
       "     returns the names of the estimated effects (coefficents) of the\n",
       "     model; ‘removeResults’ returns a ‘DESeqDataSet’ object with\n",
       "     results columns removed.\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     results(\n",
       "       object,\n",
       "       contrast,\n",
       "       name,\n",
       "       lfcThreshold = 0,\n",
       "       altHypothesis = c(\"greaterAbs\", \"lessAbs\", \"greater\", \"less\"),\n",
       "       listValues = c(1, -1),\n",
       "       cooksCutoff,\n",
       "       independentFiltering = TRUE,\n",
       "       alpha = 0.1,\n",
       "       filter,\n",
       "       theta,\n",
       "       pAdjustMethod = \"BH\",\n",
       "       filterFun,\n",
       "       format = c(\"DataFrame\", \"GRanges\", \"GRangesList\"),\n",
       "       test,\n",
       "       addMLE = FALSE,\n",
       "       tidy = FALSE,\n",
       "       parallel = FALSE,\n",
       "       BPPARAM = bpparam(),\n",
       "       minmu = 0.5\n",
       "     )\n",
       "     \n",
       "     resultsNames(object)\n",
       "     \n",
       "     removeResults(object)\n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       "  object: a DESeqDataSet, on which one of the following functions has\n",
       "          already been called: ‘DESeq’, ‘nbinomWaldTest’, or\n",
       "          ‘nbinomLRT’\n",
       "\n",
       "contrast: this argument specifies what comparison to extract from the\n",
       "          ‘object’ to build a results table. one of either:\n",
       "\n",
       "            • a character vector with exactly three elements: the name\n",
       "              of a factor in the design formula, the name of the\n",
       "              numerator level for the fold change, and the name of the\n",
       "              denominator level for the fold change (simplest case)\n",
       "\n",
       "            • a list of 2 character vectors: the names of the fold\n",
       "              changes for the numerator, and the names of the fold\n",
       "              changes for the denominator. these names should be\n",
       "              elements of ‘resultsNames(object)’. if the list is length\n",
       "              1, a second element is added which is the empty character\n",
       "              vector, ‘character()’. (more general case, can be to\n",
       "              combine interaction terms and main effects)\n",
       "\n",
       "            • a numeric contrast vector with one element for each\n",
       "              element in ‘resultsNames(object)’ (most general case)\n",
       "\n",
       "          If specified, the ‘name’ argument is ignored.\n",
       "\n",
       "    name: the name of the individual effect (coefficient) for building\n",
       "          a results table. Use this argument rather than ‘contrast’ for\n",
       "          continuous variables, individual effects or for individual\n",
       "          interaction terms. The value provided to ‘name’ must be an\n",
       "          element of ‘resultsNames(object)’.\n",
       "\n",
       "lfcThreshold: a non-negative value which specifies a log2 fold change\n",
       "          threshold. The default value is 0, corresponding to a test\n",
       "          that the log2 fold changes are equal to zero. The user can\n",
       "          specify the alternative hypothesis using the ‘altHypothesis’\n",
       "          argument, which defaults to testing for log2 fold changes\n",
       "          greater in absolute value than a given threshold. If\n",
       "          ‘lfcThreshold’ is specified, the results are for Wald tests,\n",
       "          and LRT p-values will be overwritten.\n",
       "\n",
       "altHypothesis: character which specifies the alternative hypothesis,\n",
       "          i.e. those values of log2 fold change which the user is\n",
       "          interested in finding. The complement of this set of values\n",
       "          is the null hypothesis which will be tested. If the log2 fold\n",
       "          change specified by ‘name’ or by ‘contrast’ is written as\n",
       "          beta , then the possible values for ‘altHypothesis’ represent\n",
       "          the following alternate hypotheses:\n",
       "\n",
       "            • greaterAbs: |beta| > lfcThreshold , and p-values are\n",
       "              two-tailed\n",
       "\n",
       "            • lessAbs: |beta| < lfcThreshold , p-values are the maximum\n",
       "              of the upper and lower tests. The Wald statistic given is\n",
       "              positive, an SE-scaled distance from the closest boundary\n",
       "\n",
       "            • greater: beta > lfcThreshold\n",
       "\n",
       "            • less: beta < -lfcThreshold\n",
       "\n",
       "listValues: only used if a list is provided to ‘contrast’: a numeric of\n",
       "          length two: the log2 fold changes in the list are multiplied\n",
       "          by these values. the first number should be positive and the\n",
       "          second negative.  by default this is ‘c(1,-1)’\n",
       "\n",
       "cooksCutoff: theshold on Cook's distance, such that if one or more\n",
       "          samples for a row have a distance higher, the p-value for the\n",
       "          row is set to NA. The default cutoff is the .99 quantile of\n",
       "          the F(p, m-p) distribution, where p is the number of\n",
       "          coefficients being fitted and m is the number of samples. Set\n",
       "          to ‘Inf’ or ‘FALSE’ to disable the resetting of p-values to\n",
       "          NA. Note: this test excludes the Cook's distance of samples\n",
       "          belonging to experimental groups with only 2 samples.\n",
       "\n",
       "independentFiltering: logical, whether independent filtering should be\n",
       "          applied automatically\n",
       "\n",
       "   alpha: the significance cutoff used for optimizing the independent\n",
       "          filtering (by default 0.1). If the adjusted p-value cutoff\n",
       "          (FDR) will be a value other than 0.1, ‘alpha’ should be set\n",
       "          to that value.\n",
       "\n",
       "  filter: the vector of filter statistics over which the independent\n",
       "          filtering will be optimized. By default the mean of\n",
       "          normalized counts is used.\n",
       "\n",
       "   theta: the quantiles at which to assess the number of rejections\n",
       "          from independent filtering\n",
       "\n",
       "pAdjustMethod: the method to use for adjusting p-values, see\n",
       "          ‘?p.adjust’\n",
       "\n",
       "filterFun: an optional custom function for performing independent\n",
       "          filtering and p-value adjustment, with arguments ‘res’ (a\n",
       "          DESeqResults object), ‘filter’ (the quantitity for filtering\n",
       "          tests), ‘alpha’ (the target FDR), ‘pAdjustMethod’. This\n",
       "          function should return a DESeqResults object with a ‘padj’\n",
       "          column.\n",
       "\n",
       "  format: character, either ‘\"DataFrame\"’, ‘\"GRanges\"’, or\n",
       "          ‘\"GRangesList\"’, whether the results should be printed as a\n",
       "          ‘DESeqResults’ DataFrame, or if the results DataFrame should\n",
       "          be attached as metadata columns to the ‘GRanges’ or\n",
       "          ‘GRangesList’ ‘rowRanges’ of the ‘DESeqDataSet’. If the\n",
       "          ‘rowRanges’ is a ‘GRangesList’, and ‘GRanges’ is requested,\n",
       "          the range of each gene will be returned\n",
       "\n",
       "    test: this is automatically detected internally if not provided.\n",
       "          the one exception is after ‘nbinomLRT’ has been run,\n",
       "          ‘test=\"Wald\"’ will generate Wald statistics and Wald test\n",
       "          p-values.\n",
       "\n",
       "  addMLE: if ‘betaPrior=TRUE’ was used (non-default), this logical\n",
       "          argument specifies if the \"unshrunken\" maximum likelihood\n",
       "          estimates (MLE) of log2 fold change should be added as a\n",
       "          column to the results table (default is FALSE). This argument\n",
       "          is preserved for backward compatability, as now\n",
       "          ‘betaPrior=TRUE’ by default and the recommended pipeline is\n",
       "          to generate shrunken MAP estimates using ‘lfcShrink’. This\n",
       "          argument functionality is only implemented for ‘contrast’\n",
       "          specified as three element character vectors.\n",
       "\n",
       "    tidy: whether to output the results table with rownames as a first\n",
       "          column 'row'. the table will also be coerced to ‘data.frame’\n",
       "\n",
       "parallel: if FALSE, no parallelization. if TRUE, parallel execution\n",
       "          using ‘BiocParallel’, see next argument ‘BPPARAM’\n",
       "\n",
       " BPPARAM: an optional parameter object passed internally to ‘bplapply’\n",
       "          when ‘parallel=TRUE’. If not specified, the parameters last\n",
       "          registered with ‘register’ will be used.\n",
       "\n",
       "   minmu: lower bound on the estimated count (used when calculating\n",
       "          contrasts)\n",
       "\n",
       "_\bD_\be_\bt_\ba_\bi_\bl_\bs:\n",
       "\n",
       "     The results table when printed will provide the information about\n",
       "     the comparison, e.g. \"log2 fold change (MAP): condition treated vs\n",
       "     untreated\", meaning that the estimates are of log2(treated /\n",
       "     untreated), as would be returned by\n",
       "     ‘contrast=c(\"condition\",\"treated\",\"untreated\")’. Multiple results\n",
       "     can be returned for analyses beyond a simple two group comparison,\n",
       "     so ‘results’ takes arguments ‘contrast’ and ‘name’ to help the\n",
       "     user pick out the comparisons of interest for printing a results\n",
       "     table. The use of the ‘contrast’ argument is recommended for exact\n",
       "     specification of the levels which should be compared and their\n",
       "     order.\n",
       "\n",
       "     If ‘results’ is run without specifying ‘contrast’ or ‘name’, it\n",
       "     will return the comparison of the last level of the last variable\n",
       "     in the design formula over the first level of this variable. For\n",
       "     example, for a simple two-group comparison, this would return the\n",
       "     log2 fold changes of the second group over the first group (the\n",
       "     reference level). Please see examples below and in the vignette.\n",
       "\n",
       "     The argument ‘contrast’ can be used to generate results tables for\n",
       "     any comparison of interest, for example, the log2 fold change\n",
       "     between two levels of a factor, and its usage is described below.\n",
       "     It can also accomodate more complicated numeric comparisons. Note\n",
       "     that ‘contrast’ will set to 0 the estimated LFC in a comparison of\n",
       "     two groups, where all of the counts in the two groups are equal to\n",
       "     0 (while other groups have positive counts), while ‘name’ will not\n",
       "     automatically set these LFC to 0. The test statistic used for a\n",
       "     contrast is:\n",
       "\n",
       "                        c' beta / sqrt( c' Sigma c )                    \n",
       "     \n",
       "     The argument ‘name’ can be used to generate results tables for\n",
       "     individual effects, which must be individual elements of\n",
       "     ‘resultsNames(object)’. These individual effects could represent\n",
       "     continuous covariates, effects for individual levels, or\n",
       "     individual interaction effects.\n",
       "\n",
       "     Information on the comparison which was used to build the results\n",
       "     table, and the statistical test which was used for p-values (Wald\n",
       "     test or likelihood ratio test) is stored within the object\n",
       "     returned by ‘results’. This information is in the metadata columns\n",
       "     of the results table, which is accessible by calling ‘mcols’ on\n",
       "     the ‘DESeqResults’ object returned by ‘results’.\n",
       "\n",
       "     On p-values:\n",
       "\n",
       "     By default, independent filtering is performed to select a set of\n",
       "     genes for multiple test correction which maximizes the number of\n",
       "     adjusted p-values less than a given critical value ‘alpha’ (by\n",
       "     default 0.1). See the reference in this man page for details on\n",
       "     independent filtering. The filter used for maximizing the number\n",
       "     of rejections is the mean of normalized counts for all samples in\n",
       "     the dataset. Several arguments from the ‘filtered_p’ function of\n",
       "     the genefilter package (used within the ‘results’ function) are\n",
       "     provided here to control the independent filtering behavior. In\n",
       "     DESeq2 version >= 1.10, the threshold that is chosen is the lowest\n",
       "     quantile of the filter for which the number of rejections is close\n",
       "     to the peak of a curve fit to the number of rejections over the\n",
       "     filter quantiles. 'Close to' is defined as within 1 residual\n",
       "     standard deviation. The adjusted p-values for the genes which do\n",
       "     not pass the filter threshold are set to ‘NA’.\n",
       "\n",
       "     By default, ‘results’ assigns a p-value of ‘NA’ to genes\n",
       "     containing count outliers, as identified using Cook's distance.\n",
       "     See the ‘cooksCutoff’ argument for control of this behavior.\n",
       "     Cook's distances for each sample are accessible as a matrix\n",
       "     \"cooks\" stored in the ‘assays()’ list. This measure is useful for\n",
       "     identifying rows where the observed counts might not fit to a\n",
       "     Negative Binomial distribution.\n",
       "\n",
       "     For analyses using the likelihood ratio test (using ‘nbinomLRT’),\n",
       "     the p-values are determined solely by the difference in deviance\n",
       "     between the full and reduced model formula. A single log2 fold\n",
       "     change is printed in the results table for consistency with other\n",
       "     results table outputs, however the test statistic and p-values may\n",
       "     nevertheless involve the testing of one or more log2 fold changes.\n",
       "     Which log2 fold change is printed in the results table can be\n",
       "     controlled using the ‘name’ argument, or by default this will be\n",
       "     the estimated coefficient for the last element of\n",
       "     ‘resultsNames(object)’.\n",
       "\n",
       "     If ‘useT=TRUE’ was specified when running ‘DESeq’ or\n",
       "     ‘nbinomWaldTest’, then the p-value generated by ‘results’ will\n",
       "     also make use of the t distribution for the Wald statistic, using\n",
       "     the degrees of freedom in ‘mcols(object)$tDegreesFreedom’.\n",
       "\n",
       "_\bV_\ba_\bl_\bu_\be:\n",
       "\n",
       "     For ‘results’: a ‘DESeqResults’ object, which is a simple subclass\n",
       "     of DataFrame. This object contains the results columns:\n",
       "     ‘baseMean’, ‘log2FoldChange’, ‘lfcSE’, ‘stat’, ‘pvalue’ and\n",
       "     ‘padj’, and also includes metadata columns of variable\n",
       "     information. The ‘lfcSE’ gives the standard error of the\n",
       "     ‘log2FoldChange’. For the Wald test, ‘stat’ is the Wald statistic:\n",
       "     the ‘log2FoldChange’ divided by ‘lfcSE’, which is compared to a\n",
       "     standard Normal distribution to generate a two-tailed ‘pvalue’.\n",
       "     For the likelihood ratio test (LRT), ‘stat’ is the difference in\n",
       "     deviance between the reduced model and the full model, which is\n",
       "     compared to a chi-squared distribution to generate a ‘pvalue’.\n",
       "\n",
       "     For ‘resultsNames’: the names of the columns available as results,\n",
       "     usually a combination of the variable name and a level\n",
       "\n",
       "     For ‘removeResults’: the original ‘DESeqDataSet’ with results\n",
       "     metadata columns removed\n",
       "\n",
       "_\bR_\be_\bf_\be_\br_\be_\bn_\bc_\be_\bs:\n",
       "\n",
       "     Richard Bourgon, Robert Gentleman, Wolfgang Huber: Independent\n",
       "     filtering increases detection power for high-throughput\n",
       "     experiments. PNAS (2010), <URL:\n",
       "     http://dx.doi.org/10.1073/pnas.0914005107>\n",
       "\n",
       "_\bS_\be_\be _\bA_\bl_\bs_\bo:\n",
       "\n",
       "     ‘DESeq’, ‘lfcShrink’\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "     ## Example 1: two-group comparison\n",
       "     \n",
       "     dds <- makeExampleDESeqDataSet(m=4)\n",
       "     \n",
       "     dds <- DESeq(dds)\n",
       "     res <- results(dds, contrast=c(\"condition\",\"B\",\"A\"))\n",
       "     \n",
       "     # with more than two groups, the call would look similar, e.g.:\n",
       "     # results(dds, contrast=c(\"condition\",\"C\",\"A\"))\n",
       "     # etc.\n",
       "     \n",
       "     ## Example 2: two conditions, two genotypes, with an interaction term\n",
       "     \n",
       "     dds <- makeExampleDESeqDataSet(n=100,m=12)\n",
       "     dds$genotype <- factor(rep(rep(c(\"I\",\"II\"),each=3),2))\n",
       "     \n",
       "     design(dds) <- ~ genotype + condition + genotype:condition\n",
       "     dds <- DESeq(dds) \n",
       "     resultsNames(dds)\n",
       "     \n",
       "     # the condition effect for genotype I (the main effect)\n",
       "     results(dds, contrast=c(\"condition\",\"B\",\"A\"))\n",
       "     \n",
       "     # the condition effect for genotype II\n",
       "     # this is, by definition, the main effect *plus* the interaction term\n",
       "     # (the extra condition effect in genotype II compared to genotype I).\n",
       "     results(dds, list( c(\"condition_B_vs_A\",\"genotypeII.conditionB\") ))\n",
       "     \n",
       "     # the interaction term, answering: is the condition effect *different* across genotypes?\n",
       "     results(dds, name=\"genotypeII.conditionB\")\n",
       "      \n",
       "     ## Example 3: two conditions, three genotypes\n",
       "     \n",
       "     # ~~~ Using interaction terms ~~~\n",
       "     \n",
       "     dds <- makeExampleDESeqDataSet(n=100,m=18)\n",
       "     dds$genotype <- factor(rep(rep(c(\"I\",\"II\",\"III\"),each=3),2))\n",
       "     design(dds) <- ~ genotype + condition + genotype:condition\n",
       "     dds <- DESeq(dds)\n",
       "     resultsNames(dds)\n",
       "     \n",
       "     # the condition effect for genotype I (the main effect)\n",
       "     results(dds, contrast=c(\"condition\",\"B\",\"A\"))\n",
       "     \n",
       "     # the condition effect for genotype III.\n",
       "     # this is the main effect *plus* the interaction term\n",
       "     # (the extra condition effect in genotype III compared to genotype I).\n",
       "     results(dds, contrast=list( c(\"condition_B_vs_A\",\"genotypeIII.conditionB\") ))\n",
       "      \n",
       "     # the interaction term for condition effect in genotype III vs genotype I.\n",
       "     # this tests if the condition effect is different in III compared to I\n",
       "     results(dds, name=\"genotypeIII.conditionB\")\n",
       "     \n",
       "     # the interaction term for condition effect in genotype III vs genotype II.\n",
       "     # this tests if the condition effect is different in III compared to II\n",
       "     results(dds, contrast=list(\"genotypeIII.conditionB\", \"genotypeII.conditionB\"))\n",
       "     \n",
       "     # Note that a likelihood ratio could be used to test if there are any\n",
       "     # differences in the condition effect between the three genotypes.\n",
       "     \n",
       "     # ~~~ Using a grouping variable ~~~\n",
       "     \n",
       "     # This is a useful construction when users just want to compare\n",
       "     # specific groups which are combinations of variables.\n",
       "     \n",
       "     dds$group <- factor(paste0(dds$genotype, dds$condition))\n",
       "     design(dds) <- ~ group\n",
       "     dds <- DESeq(dds)\n",
       "     resultsNames(dds)\n",
       "     \n",
       "     # the condition effect for genotypeIII\n",
       "     results(dds, contrast=c(\"group\", \"IIIB\", \"IIIA\"))\n",
       "     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "?results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "funky-convert",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
