---
title: "Analysis of differential expression across conditions"
output: html_notebook
---



```{r}
library("DESeq2")
setwd("C:/users/tinta/OneDrive/Documents/Projects/Proyecto_rodopsina/")
load("Data/DokdoniaTPMCounts.RData")

cnames <- colnames(data$counts)
conditions = unlist(
    lapply(cnames, function(s){return(strsplit(s, ".sam")[[1]])})
)

light_conditions <- conditions[1:16]
dark_conditions <- conditions[17:length(conditions)]

light <- vector()
temp <- vector()
files <- colnames(data$counts)
for (i in 1:length(files)) {
    light <- c(light, substring(files[i], 1, 1))
    temp <- c(temp, substring(files[i], 3, 4))
}

meta <- data.frame(row.names = cnames,
                  "Temperature" = temp,
                  "Light" = light)
head(meta, 5)
```


```{r}
dds <- DESeqDataSetFromMatrix(countData = data$counts, 
                              colData = meta, design = ~ Temperature)  
dds_lrt <- DESeq(dds, test="LRT", reduced = ~ 1) # Likelyhood Ratio Test (multiple conditions)

res <- results(dds_lrt)
res
```

Negative log2 gold chnges impy a negative correlation of gene expression with temperature.
Let's filter these results.

```{r}
res$log2FoldChange

```

Let's plot the temerpature series to visualize the correlation

```{python}
import pandas as pd
tpm = pd.read_csv('C:/users/tinta/OneDrive/Documents/Projects/Proyecto_rodopsina/Data/tpm_counts.csv', index_col=0)
gene_id = 'MED134_14802'
tpm.loc[gene_id].plot(legend=False, figsize=(15, 10))
```

